lib/puppet/pops/types/types.rb

```diff
@@ -736,7 +739,7 @@ def self.register_ptype(loader, ir)
   end
 
   def instance?(o, guard = nil)
-    return o.is_a?(String) || o.is_a?(Integer) || o.is_a?(Float) || o.is_a?(TrueClass) || o.is_a?(FalseClass)
+    return o.instance_of?(String) || o.is_a?(Integer) || o.is_a?(Float) || o.instance_of?(TrueClass) || o.instance_of?(FalseClass)
```

Comment:
I'd lean towards keeping it as-is for consistency with existing code like https://github.com/puppetlabs/puppet/blob/1156ac702c369930434bed2cedacdfe3a30a50e9/lib/puppet/pops/types/types.rb#L988

But feel free to fix both of those in a separate commit or JIRA ticket even? Note there's some `Numeric`/`Float`/`Rational` confusion throughout the pops types code. For example, the type_calculator doesn't special case `Float` in https://github.com/puppetlabs/puppet/blob/1156ac702c369930434bed2cedacdfe3a30a50e9/lib/puppet/pops/types/type_calculator.rb#L257 That optimization is done in cases where puppet needs to infer if an object has a callable method like:

```puppet
notice(1.2.abs)
```

There's also some confusion around `Numeric` and `Rational`. Commit d411dc381fafe08c64b980e548050902850bd120 tried to ensure `Rational` objects have type `PRuntimeType`. However, the type calculator considers them to be instances of `PScalarType` and `PRuntimeType`

```
$ bx irb
irb(main):001:0> require 'puppet'
=> true
irb(main):002:0> calc = Puppet::Pops::Types::TypeCalculator.new
=> #<Puppet::Pops::Types::TypeCalculator:0x0000558cffad5778 @infer_visitor=#<Puppet::Pops::Visitor:0x0000558cffad5728 @receiver=nil, @message="infer", @min_args=0, @max_args=0, @cache={}>, @extract_visitor=#<Puppet::Pops::Visitor:0x0000558cffad56b0 @receiver=nil, @message="extract", @min_args=0, @max_args=0, @cache={}>>
irb(main):003:0> calc.instance?(Puppet::Pops::Types::TypeFactory.scalar, Rational(0))
=> true
irb(main):004:0> calc.instance?(Puppet::Pops::Types::TypeFactory.runtime, Rational(0))
=> true
```

I think this is a bug in:

https://github.com/puppetlabs/puppet/blob/1156ac702c369930434bed2cedacdfe3a30a50e9/lib/puppet/pops/types/types.rb#L697

Since this is unrelated to the original regression, I'd file it as a separate issue